cranelift: Static check out of bounds accesses to `heap_addr`'s

afonso360 opened the pull request:
👋 Hey,

I don't know if we should merge this or not!

This PR started on https://github.com/bytecodealliance/wasmtime/pull/5155#discussion_r1009872167 where @jameysharp pointed out that it was a bit weird that the interpreter used a `saturating_sub` when calculating `heap_addr` sizes.

I added that because we do have some 0 sized `heap_addr`'s in runtests. However when fixing that, I added a verifier rule, and it turns out we have a **lot** of 0 sized `heap_addr`'s in our test suite. Are these legal?

The [instruction docs](https://docs.rs/cranelift-codegen/latest/cranelift_codegen/ir/trait.InstBuilder.html#method.heap_addr) state the following:
> Verify that the offset range `p .. p + Size - 1` is in bounds for the heap H, and generate an absolute address that is safe to dereference.

So it sounds like `Size` must always be larger than 1?

But it also says:

> If `p + Size` is not greater than the heap bound

Which would allow 0 as a `Size`.

So, lets make a decision about this! I don't have any preference for either, I'm just opening this as a PR because when I started fixing the interpreter I thought it would just be that one runtest that was wrong!

---

Also, there are some regressions in the x64 codegen that probably shouldn't be there.

cfallin commented:
Really interesting question here!

My first impulse is actually slightly different: in general I don't like edge cases or discontinuities, and for the same reason that I like, e.g., zero-sized slices to be possible, I think that a `heap_addr` with a size of `0` should be legal. It should execute without a trap if the base address given is less than *or equal to* the length of the heap.

In other words, the `base + size - 1` computation finds the address of the last byte of the thing being accessed, and this implicitly assumes there *is* a last byte (i.e. it's like a `vec.last().unwrap()`). Instead of `base + size - 1 < heap.base + heap.len`, I might suggest `base + size <= heap.base + heap.len`. That naturally handles the zero-sized case, and feels like the right answer for "limit of behavior as size approaches zero". (In general I'm also slightly allergic to `- 1` and `+ 1` in offset computations; they feel like a sign that we haven't found the most natural expression of the computation. For the same reasons that 0-based array indexing makes more sense etc etc)

FWIW, the instruction docs just below the quoted part do seem consistent with this: "If `p + Size` is greater than the heap bound, generate a trap." (Inverting "greater than" we get "less than or equal", and "heap bound" is `heap.base + heap.len`.)

Of course it's invalid to actually *use* the computed pointer to do any access of size greater than zero (that is, any access at all); but that's perfectly fine. That does mean that the tests that you're correcting here are illegal anyway, as they do `v + 8` and `v + 128` on the resulting pointers. So separately, we should update those tests to use appropriate `Size`s (though it doesn't actually matter for the properties we're testing).

Thoughts?

afonso360 commented:
That sounds reasonable! 

And I think we can enforce some of that via the verifier. If we can compute all of the offsets, and that a `load`/`store`  is directly referencing a `heap_addr`,  we can probably check if the access is legal. This still misses a bunch of cases, but I think it might catch the most common ones.

I'm going to give this a try.

afonso360 commented:
Added the above, turns out our testsuite is full of out of bounds loads and stores 🙃, even more than the zero sized accesses. 

I've reverted the cases where we do a heap_addr of 0 bytes but then never use that pointer. 

One thing that jumped at me was that some of our optimization test cases regressed when we used a non 0 byte heap_addr. [This one in particular](https://github.com/bytecodealliance/wasmtime/blob/cd53bed898fa026b5b2318ed5f66599207da83b8/cranelift/filetests/filetests/isa/x64/heap-no-spectre.clif#L64-L75) I think shouldn't have regressed. We may want to follow up on that.

jameysharp commented on cranelift/codegen/src/verifier/mod.rs:
```diff
@@ -1124,6 +1148,66 @@ impl<'a> Verifier<'a> {
+        let valid_size = match dfg.value_def(addr) {
+            ValueDef::Result(def_inst, _) => match dfg[def_inst] {
+                ir::InstructionData::HeapAddr { imm, .. } => u32::from(imm) as i64,
+                // Not a `heap_addr`, we can't verify this.
```
In principle I suppose we could verify arithmetic where one operand is a `heap_addr` and the rest are all constants. I'm not suggesting that should be in this PR (unless you really want to), but maybe note it in this comment?

jameysharp commented on cranelift/codegen/src/verifier/mod.rs:
```diff
@@ -1124,6 +1148,66 @@ impl<'a> Verifier<'a> {
+            | Opcode::Sload8x8
+            | Opcode::Sload16x4
+            | Opcode::Sload32x2 => 8,
+            _ => dfg.ctrl_typevar(inst).bytes() as i64,
```
I'd lean toward explicitly listing the remaining opcodes, and having an `_ => unreachable!()` at the end. Someone who adds a new instruction that uses one of the `Load` or `Store` instruction formats may not realize they should update this `match` in the middle of the verifier.

jameysharp commented on cranelift/codegen/src/verifier/mod.rs:
```diff
@@ -1124,6 +1148,66 @@ impl<'a> Verifier<'a> {
+        }
+
+        // Check if the access is within the bounds provided by  heap_addr
+        let oob_bytes = -(valid_size - (offset + access_size));
```
This is algebraically equivalent to what you wrote, but makes more sense to me:

```suggestion
        let needed_size = offset + access_size;
        let oob_bytes = needed_size - valid_size;
```

jameysharp reviewed (approved):
I think this is great! I have a couple of suggestions but I'd merge this as-is if you want. I do want to know if @cfallin has any concerns about adding this to the validator.

cfallin commented:
I actually have a slightly different opinion here: we should not make IR validation contingent on static analysis of offsets. This is a separation-of-concerns issue in my mind: the allowed IR should be defined separately from how powerful we can make our analyses (or in other words, IR should not be disallowed because "the compiler can't prove X"). That is a recipe for very confusing issues where validation state flips back and forth as optimizations change code.

For a very concrete example, removing redundant blockparams could allow an analysis to see further; it could then suddenly observe a statically-known-to-be-invalid combinatino of size and offset; and so we have an optimization pass that takes legal IR and produces illegal IR, according to our validator.

In general, the legality of loads/stores is a *dynamic* property, not a static (type-like) property, and for consistency we should keep it that way, I think; anything else causes confusing and unintuitive behavior.

afonso360 commented:
> For a very concrete example, removing redundant blockparams could allow an analysis to see further; it could then suddenly observe a statically-known-to-be-invalid combinatino of size and offset; and so we have an optimization pass that takes legal IR and produces illegal IR, according to our validator.

Explained that way, I totally agree!

It still feels like we should be able to provide these sorts of checks to IR producers. What would you think about adding this behind a separate flag that tries to do more advanced checks in general?

cfallin commented:
> It still feels like we should be able to provide these sorts of checks to IR producers. What would you think about instead adding this behind a separate flag that tries to do more advanced checks in general?

Possibly, yeah; a "lint" level of warning, operating sort of like classical static analysis tools, could be useful. As long as we state that the results are not hard errors (i.e., are not part of the definition of "correct IR"), that seems fine to me.

afonso360 commented:
We discussed this PR yesterday in the Cranelift weekly meeting and reached the conclusion that we probably get the same coverage as this by adding heaps to fuzzgen and checking accesses in the interpreter (which we already do!).
